### DISCUSSION TITLE
Sneak peek into the future of GitHub and AI - Powered by GitHub Copilot X

### DISCUSSION THREAD

**Discussion by Akash1134** [community/community#50727] [Copilot News and Announcements]
Today, we are announcing our vision for AI, powered by [GitHub Copilot X](https://www.github.com/features/preview/copilot-x). 

A new generation of GitHub Copilot will support pull requests, answer your questions using Copilot for Docs, provide in-depth analysis of your code using Chatbot right into your editor, and much more by creating an AI assistant readily available to you. 

Check out this [blog post](https://github.blog/2023-03-22-github-copilot-x-the-ai-powered-developer-experience/) by GitHub CEO, [Thomas Dohmke](https://github.com/ashtom), where he shares a peek into the future of GitHub and AI.

![Screenshot 2023-03-22 at 7 42 26 PM](https://user-images.githubusercontent.com/95834130/226931354-341cd9c0-0d9f-4d7d-96c8-4abef1948e83.png)

Generative AI models will drive the next wave of innovation—and we want that innovation to start with developers. 

We started our mission with GitHub Copilot, which has transformed how millions of developers build software. But that’s only one piece of software development, there are manifold opportunities to integrate AI into a developer’s daily experience to optimize productivity, minimize mundane tasks, and ultimately increase happiness. [Get GitHub Copilot](https://github.com/features/copilot) today to be a part of the next generation of AI.

Resources to keep you up to date:

- [GitHub Copilot for Docs](https://www.github.com/features/preview/copilot-x#docs)
- [GitHub Copilot for PR](https://www.github.com/features/preview/copilot-x#pullrequests)(Technical Preview)
- [GitHub Copilot Chat](https://www.github.com/features/preview/copilot-x#chat)
- [GitHub Copilot for CLI](https://www.github.com/features/preview/copilot-x#cli) (Technical Preview)
- [Explore all Previews](https://github.com/features/preview) (Be the first to try out GitHub’s new features)
- [GitHub Next](https://githubnext.com/projects/hey-github/)

Let's build from here 🚀

**Comment by Neodevils** [comment]
I'm excited about GitHub Copilot chat. 👀

**Reply by Ozodbek4252** [reply]
Can regular copilot members access copilot x?

**Reply by Akash1134** [reply]
Hi @Ozodbek4252 GitHub Copilot X is currently a representation of GitHub’s vision for the future rather than an available product offering of GitHub Copilot. As we continue to design, test, and build features that fall into the GitHub Copilot X vision, we are also taking the time to determine the best way to provide them to our customers.

You can explore the Copilot X features we offer to preview [here](https://github.com/features/preview/copilot-x).

**Reply by DanDevel** [reply]
GitHub Copilot is an AI-powered coding assistant that helps developers write code faster and more efficiently. It uses machine learning to provide real-time code suggestions and auto-completions, saving time and reducing repetitive tasks. With Copilot, developers can boost productivity and improve code quality by leveraging its vast knowledge base and contextual understanding of the code being written.

**Comment by MarvinJWendt** [comment]
Hi, is there any information on how the waiting lists are processed. Is it first come, first served?

**Reply by ettaboyle** [reply]
Hi @MarvinJWendt - great question! We expect people will be curious about this.

While we are unable to guarantee access to each feature, you can submit your interest by joining the waitlist to preview the features that you are interested in. **Joining one waitlist does not confirm your interest in more than one feature, so you must join each separately**. 
- Please note we currently don’t have guaranteed timelines for access. 
- Once you get access to preview a feature, please keep in mind that what you are getting access to is considered beta or a technical preview, meaning the feature may still have some kinks to work out. 

Thank you!

**Reply by MarvinJWendt** [reply]
Hi @ettaboyle, thanks for fast reply! I understand that you don't want to give out fixed ETAs (I wouldn't do that too), but is there a very rough estimate when the roll-out to waitlist users will start? Does the announcement mean "we will start to give out access as of today", or more like "hey look what we made, we will start to grant access in a couple of months"? I remember that some of my friends waited close to a year for access to Codespaces. I am just curious if I need to check my emails daily, or if I can lean back a bit :)

**Reply by ryanjsalva** [reply]
@MarvinJWendt the honest truth is that the answer is different for each wait list. Some capabilities are ship-ready, while others need a bit of bake time before we ask developers to provide feedback. But for all capabilities, we're starting to admit a limited number of users within weeks, not months.

**Reply by inderjotpujara** [reply]
@ryanjsalva I am waitlisted on the copilot chat feature as of now 
Can you share any time (maybe in days, weeks, or months )?
would be really helpful.

**Reply by hra42** [reply]
> @ryanjsalva I am waitlisted on the copilot chat feature as of now 
> Can you share any time (maybe in days, weeks, or months )?
> would be really helpful.
> 

I doubt they will answer that question. It depends on so many factors, that it will be impossible to even give a time frame.

**Reply by bleachkitty** [reply]
I joined the waitlist about a month ago and haven't got access.  A student of mine joined less than a week ago and just got access tonight, so it's not first-come, first-serve.  I'm pretty sure it's literally random, which sucks because it's one of the main reasons I initially subscribed to copilot.

**Comment by tmlewin** [comment]
Would the copilot for individuals undergo an update in terms of gpt model . like would it support gpt4 or do we need copilot x for that?

**Reply by ettaboyle** [reply]
Hi @tmlewin - thanks for your question! I reached out to the Copilot team and they shared that this is unknown at this time. Sorry we can't provide a more definitive answer right now.

The Copilot team is paying attention to the feedback, feature requests & bugs we are hearing from users in the Community Discussions so please continue to share your thoughts. 
- If you believe this should be a feature request or enhancement please [add a discussion on the topic to the Copilot Category](https://github.com/orgs/community/discussions/new?category=copilot) and include as much detail as you can to help inform the product strategy. 
- **Tips for posting in discussions:** 1. Make sure you select the `product feedback` topic category. 2. Add a clear title related to the request. 3. If its a feature request add the `enhancement` label

Thank you!

![Screenshot 2023-03-22 at 2 24 25 PM](https://user-images.githubusercontent.com/75342807/227041914-f00c6c52-8e3b-4d3a-9f87-77725160a1d3.png)

**Reply by Akash1134** [reply]
Hi @tmlewin Thank you for asking a great question. I'd like to give you more details on your question from our engineers: Copilot for Individuals will be upgraded to new models based on quality metrics. Currently, we balance suggestion quality as measured by predefined parameters and Codex delivers the best blend for them. Hope that helps!

**Reply by ettaboyle** [reply]
Thanks @Akash1134 ✨ @tmlewin we shared your question with the product team and have more to share:

> Copilot for Individuals ("auto-complete") will be upgraded to new models based on quality metrics. We typically balance  suggestion quality as measured by:
> - number of suggestions in a session
> - Suggestion acceptance rate
> - Latency between keystroke and suggestion
> - Qualitative feedback from developers on perceived value
> Today, Codex delivers the best blend of these four metrics, but GPT 3.5-turbo is showing a lot of promise as a possible replacement.

Hopefully that helps answer your question! Thanks again

**Reply by tmlewin** [reply]
Thanks for the feedback.

**Comment by Qiangbo100** [comment]
Hi, I am interested in using GitHub Copilot X and I noticed that it is currently compatible with VS Code and Visual Studio. However, I was wondering if there are any plans to make it compatible with JetBrains IDEs in the future?

**Reply by Akash1134** [reply]
Hi @Qiangbo100 ,

Great Question! GitHub Copilot X is currently a representation of GitHub’s vision for the future rather than an available product offering of GitHub Copilot. As we continue to design, test, and build features that fall into the GitHub Copilot X vision, we are also taking the time to determine the best way to provide them to our customers.

As we work towards making it more accessible and compatible with other IDEs, we'll share that information accordingly.

**Comment by ovidb** [comment]
Can someone answer why does signup up to the waitlist requires so many permissions?

![image](https://user-images.githubusercontent.com/1531356/227165017-efcd2948-0b29-40a4-8c3b-adf1d7a646c6.png)


I would like to know why is it needed to
1. see the resources I have access too?
but most importantly
2. Why do you need to act on my behalf?
that seems very excessive and I'm having trouble convincing allowing my team to use/signup for this app until we get an explanation.

Thank you

**Reply by Akash1134** [reply]
Hi @ovidb ,

Thank you for your question. Please read the [GitHub Next Pre-Release License Terms](https://github.com/githubnext/githubnext/blob/main/TERMS_AND_CONDITIONS.md) and I hope that helps answer your questions. Feel free to get back in case you need more details. 🙂

**Reply by Akash1134** [reply]
For more details on what all those permissions are authorized for, please see [Authorizing GitHub Apps](https://docs.github.com/en/apps/using-github-apps/authorizing-github-apps)

**Reply by ovidb** [reply]
Thank you @Akash1134 

Neither of those links really address the question.

I feel those permissions a bit excessive without a proper explanation of why you specifically need to see **all** the resources I have access to and furthermore **Act on my behalf**. 

I understand you might never act on my behalf, but you could. Effectively you can **see and do whatever my user has access to see and do**. We would feel more comfortable if for example we could enrol a specific repository into the programme, but *everything* seems a bit too much

We feel a bit uncomfortable to essentially allow an app to be able to impersonate every single user in our organisation that signs-up for this. Your License Terms do not really address the issue as I previously said

We would really love to use your product and we are very excited to jump on board, but we need an explanation of why such broad net of permissions is required to use the product.

Thank you

**Reply by Akash1134** [reply]
Thank you for getting back @ovidb We have shared your feedback with the concerned team.

**Reply by ovidb** [reply]
@Akash1134 any feedback from the team?

**Comment by adga-msft** [comment]
I already got access to CoPilot X through my org account "adga_microsoft", but how do I enable the "X" version in VS Code? **I just see the normal CoPilot working in my VS code?**

<img width="361" alt="image" src="https://user-images.githubusercontent.com/127955637/227557739-17bf43b9-54b6-451e-8d49-ec26cf152071.png">

**Reply by adga-msft** [reply]
@Akash1134

**Reply by ettaboyle** [reply]
Hi @adga-msft 👋🏻 Thanks for reaching out. It's outside of @Akash1134 's working hours so I'm looking into this 🕵🏻‍♂️ I'll share more info when I hear back from the team here.

**Reply by veverkap** [reply]
@adga-msft please make sure that you are using the latest [VS Code Insiders](https://code.visualstudio.com/insiders/) and the latest [GitHub Copilot Nightly extension](https://marketplace.visualstudio.com/items?itemName=GitHub.copilot-nightly)

**Reply by SemperFu** [reply]
Is there a nightly extension for Visual Studio?

**Reply by Akash1134** [reply]
Hi @SemperFu there is currently no nightly extension for VS. The[ one available](https://marketplace.visualstudio.com/items?itemName=GitHub.copilot-nightly) is for VS Code. For Copilot X, you'll need VS Code insiders with the latest GitHub Copilot Nightly extension Thanks!

**Reply by pealbuqu** [reply]
@Akash1134 is this the only way to use the beta version? The FAQ says that the technical preview is available for vs and vscode.
![image](https://user-images.githubusercontent.com/117676522/229635084-49e10877-59c0-4621-8375-2071e507fea4.png)

**Reply by MarvinJWendt** [reply]
@pealbuqu when you join the waitlist, you have to wait until you get the access. You'll be informed, via email, when your access is ready.

**Reply by pealbuqu** [reply]
@MarvinJWendt I have access. I think I haven't received an e-mail yet, but if I try to join the waitlist again, I get the same message as adga-msft. I followed the process that Akash1134 mentioned and now I can use the beta version in vscode, but I'm wondering if there is an alternate process to use it in visual studio instead.

**Reply by pealbuqu** [reply]
@Akash1134 is there a way to use the technical preview in visual studio? I already have access to Copilot chat.

**Reply by Akash1134** [reply]
> is there a way to use the technical preview in visual studio? I already have access to Copilot chat.


Hi @pealbuqu , Not at the moment.  For Copilot X, you'll need VS Code insiders with the latest GitHub Copilot Nightly extension Thanks!

**Comment by idobukhov** [comment]
It's really cool

**Comment by mrhobbeys** [comment]
I'm excited! If we get into the preview, what kinds of information will the devs be looking for? Do they just want to see how people will use it, or are they trying to find situations where it will give wrong answers?

**Reply by Akash1134** [reply]
Hi @mrhobbeys That's a great question. As and when you get approved for the technical preview for any of the Copilot X features, you will  be able to experience the product yourself and play around with the kind of suggestions you'll get. Honestly, it is too early for me to comment with a concrete statement. As we continue to design, test, and build features that fall into the GitHub Copilot X vision, we are also taking the time to determine the best way to provide them to our customers.

**Comment by eshack94** [comment]
How can I find out my current "position" on the waitlist for Copilot Chat? (sorry if this isn't the right place to ask)

**Reply by Akash1134** [reply]
Hi @eshack94 , 


Currently, there is no way to find out one's position on the Copilot Chat waitlist. Currently, Copilot Chat is only available for public preview. 

This is the right place to ask questions but it would be best to post them in the community as a new discussion. Copilot team is paying attention to the feedback, feature requests & bugs we are hearing from users in the Community Discussions so please continue to share your thoughts.

For future posts:

- If you believe this should be a feature request or enhancement or question please [add a discussion on the topic to the Copilot Category](https://github.com/orgs/community/discussions/new?category=copilot) and include as much detail as you can to help inform the product strategy.
- Tips for posting in discussions: 1. Make sure you select the product feedback topic category. 2. Add a clear title related to the request. 3. If its a feature request add the enhancement label or add a question if it's a query (like in your case) Thanks!

**Comment by shomos555** [comment]
I also want access to testing copilot x.((( But it seems that the waiting list is not moving. I am writing a project in flutter for smartphones, and I am very weak in programming. I have already written half of it practically using GPT 4, but it has become unbearably laggy and stupid So I would test your new copilot properly.

**Reply by shomos555** [reply]
Gave access, thanks. Tell me if something personally needs to be done with the chat in visual studio, if you need some better feedback. And then I will just work with my project, as I understand it, the model is trained on this anyway and you get the necessary data.

**Reply by JonasDev17** [reply]
If you're not good at programming then I do not recommend relying on chatgpt or copilot too much or otherwise you will never properly learn it and all you will do is mindlessly copy paste code. What you need imo is practice, not copilot. I recommend chatgpt for explaing things or helping you find bugs and then have chatgpt explain the fix to help you learn. I learned programming when all this didn't exist.

**Comment by beyondcompute** [comment]
Is the ability to simplify/refactor the existing code planned or coming? In most projects where I worked writing new code is not a problem. In fact, often creating new features is such a rare and pleasant task that developers would not too eagerly part with it. Anyway, the real problem is simplifying large amounts of code in a safe way. If AI can solve that, it would be a game-changer! 🤖

**Reply by Akash1134** [reply]
Hi @beyondcompute , GitHub Copilot X is currently a representation of GitHub’s vision for the future rather than an available product offering of GitHub Copilot. As we continue to design, test, and build features that fall into the GitHub Copilot X vision, we are also taking the time to determine the best way to provide them to our customers. 

Our R&D team at [GitHub Next](https://githubnext.com/) has been working to move past the editor and evolve GitHub Copilot into a readily accessible AI assistant throughout the entire development lifecycle. Simplifying a large amount of code is one of our long-term goals and we'll have more to share once the product develops based on your inputs.

**Comment by MovGP0** [comment]
Is it possible to link Copilot with a vector database like qdrant or pinecone, such that I can make queries against the whole codebase, instead of just the current tab(s)?

**Reply by Akash1134** [reply]
Hi @MovGP0, we don't offer that ability today but this is worth a feature request. 
- Please [add a discussion on the topic to the Copilot Category](https://github.com/orgs/community/discussions/new?category=copilot) and include as much detail as you can to help inform the product strategy.
- Tips for posting in discussions: 1. Make sure you select the product feedback topic category. 2. Add a clear title related to the request. 3. If its a feature request add the enhancement label or add a question if it's a query (like in your case) Thanks!

**Reply by MovGP0** [reply]
@Akash1134  [Recurrent Memory Transfer](https://arxiv.org/abs/2304.11062) (RMT) might be an alternative.

**Comment by wenxiayili** [comment]
Hi, does anyone know how long it takes for the GitHub Copilot Chat waitlist to be reviewed?"

**Comment by cmp-nct** [comment]
Anyone who has access (I believe the number is in the tens?) can compare it to Amazon CodeWhisperer ?
Also give it a test how many tokens the "new copilot" can see of your code, the old one was around 8k at release then 4k now 1k.
Codewhisperer is also low, just 2k tokens.

Just test carefully how much "up" it can see in your code.
You can then get the token count here: https://platform.openai.com/tokenizer

**Reply by shomos555** [reply]
They began to provide access, you need to try and check whether they gave you. I don't understand, maybe it's already available to everyone. you need to download all the prototypes of the environment according to the instructions provided above. Also, as I understand it, install the git hub extension copylot chat

**Comment by ivanv-microsoft** [comment]
![image](https://github.com/community/community/assets/28568469/19bace4f-9455-4e11-9c4d-e84eba4923fc)

Do we still need to explicitly sign up?

**Reply by ettaboyle** [reply]
Hi @ivanv-microsoft 👋🏻 This announcement was from last year (2023). I'm not sure what link you're trying to access, so that may have resulted in a 404.

> [!TIP]
> @Akash1134 is our Community Manager for the [Copilot Category](https://github.com/orgs/community/discussions/categories/copilot)  in our community discussions and put together this **[GitHub Copilot FAQs + Troubleshooting common issues ⚙️](https://github.com/orgs/community/discussions/47318)**. Hopefully that helps! 

If you're still having issues, search in the discussions to see if another user has experienced a similar issue, and if not go ahead and open up a new discussion in the appropriate category.

Thank you!

**Reply by ivanv-microsoft** [reply]
Oh, my bad. forgot to include the link. We got this from the "aka.ms/FreeThings" oneNote:

Steps for Copilot + ChatGPT:
Once you got access to regular Copilot, you can then get access to Copilot X (with ChatGPT support), you need to go here to get access: **https://github.com/github-copilot/chat_waitlist_signup/join**
Then follow these instructions for setup (VS Code Insiders and GitHub Copilot Nightly) https://github.com/orgs/community/discussions/50727#discussioncomment-5421791 (basically, you need to install insiders version of VS Code, and then install GitHub Copilot Nightly from the extensions).

**Reply by ettaboyle** [reply]
Hmm.. okay! Let me check with some folks and see what I can find out. @Akash1134 is not on right now or else I would defer to him

**Reply by ettaboyle** [reply]
@ivanv-microsoft guidance is for all MSFT employees to go to https://aka.ms/copilot

**Comment by Jv2350** [comment]
I'm particularly interested in Copilot for Docs. Writing documentation can be a pain, so this could be a huge time saver

**Comment by byKosta** [comment]
I just want to say that as an assistant "GitHub Copilot Chat" is just great. btw the name fits perfectly :slightly_smiling_face:
